psl-core/src/main/java/org/linqs/psl/reasoner/sgd/SGDReasoner.java

```diff
@@ -60,60 +88,143 @@ public double optimize(TermStore baseTermStore) {
+                oldVariableValues = Arrays.copyOf(termStore.getVariableValues(), termStore.getVariableValues().length);
+                oldObjective = Double.POSITIVE_INFINITY;
+            } else {
+                System.arraycopy(termStore.getVariableValues(), 0, oldVariableValues, 0, oldVariableValues.length);
+                oldObjective = objective;
+            }
+
+            long end = System.currentTimeMillis();
             totalTime += end - start;
 
             if (log.isTraceEnabled()) {
-                log.trace("Iteration {} -- Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimiztion Time: {}",
-                        iteration, objective, movement, (end - start), totalTime);
+                log.trace("Iteration {} -- Objective: {}, Normalized Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimization Time: {}",
+                        iteration, objective, objective / termCount, meanMovement, (end - start), totalTime);
             }
 
             iteration++;
             termStore.iterationComplete();
-
-            if (breakOptimization(iteration, objective, oldObjective, movement)) {
-                break;
-            }
         }
 
-        log.info("Optimization completed in {} iterations. Objective: {}, Total Optimiztion Time: {}",
-                iteration - 1, objective, totalTime);
-        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termStore.size());
+        objective = computeObjective(termStore);
+        log.info("Optimization completed in {} iterations. Objective: {}, Normalized Objective: {}, Total Optimization Time: {}",
+                iteration, objective, objective / termCount, totalTime);
+        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termCount);
 
         termStore.syncAtoms();
 
         return objective;
     }
 
-    private boolean breakOptimization(int iteration, double objective, double oldObjective, float movement) {
+    private float minimize(SGDObjectiveTerm term, float[] variableValues, int iteration) {
+        float movement = 0.0f;
+        float variableStep = 0.0f;
+        float newValue = 0.0f;
+        int[] variableIndexes = term.getVariableIndexes();
+
+        term.computeGradient(variableValues, gradient);
+
+        for (int i = 0; i < term.size(); i++) {
+            if (coordinateStep) {
+                term.computeGradient(variableValues, gradient);
+            }
+            variableStep = computeVariableStep(variableIndexes[i], iteration);
+
+            newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - variableStep));
+            movement += Math.abs(newValue - variableValues[variableIndexes[i]]);
+            variableValues[variableIndexes[i]] = newValue;
+        }
+
+        return movement;
+    }
+
+    private float computeVariableStep(int variableIndex, int iteration) {
```

Comment:
Hope you have tested all approaches and made sure it converges. And plan on adding unit tests to ensure they are correct.

Reply:
Ran experiments on 8 PSL examples (epinions citeseer cora jester lastfm entity-resolution simple-acquaintances friendship) and observed convergence on all of them.
Here is a google sheet of 3 of these experiments: 
https://drive.google.com/drive/folders/1MPTfYrqD_d-MspsUs6iu8CqAYPXz0i1s?usp=sharing